### PR TITLE
[AAQ-281] Multilanguage content support

### PR DESCRIPTION
**Pull request by suzinyou**
Reviewer: @amiraliemami  

Estimate: 1hr

---

## Ticket

Fixes: [AAQ-281](https://idinsight.atlassian.net/browse/AAQ-281)

### Goal
To add simple multilingual support on the backend.

### Changes
1. When creating content, we now need to specify language
2. Based on detected language of the query, we do similarity search only within contents in the same language.
3. Language is optional on the backend. If there is no detected language we search across all content.
4. + LLM response now summarizes top N responses!

> [!WARNING]
> Some of the changes we made here, e.g. 4, we might want to merge to main as well. I can cherry-pick and raise another PR later.

## How has this been tested?
- Please try all the content management and question answering APIs on dev mode
- All unit tests should pass

## Checklist

Fill with `x` for completed. Delete any lines that are not relevant

- [ ] My code follows the style guidelines of this project
- [ ] I have reviewed my own code to ensure good quality
- [ ] I have tested the functionality of my code to ensure it works as intended
- [ ] I have resolved merge conflicts
- [ ] I have updated the automated tests (if applicable)
- [ ] I have updated the requirements (if applicable)
- [ ] I have updated the README file (if applicable)
- [ ] I have updated affected documentation (if applicable)

[AAQ-281]: https://idinsight.atlassian.net/browse/AAQ-281?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: core_backend/app/configs/llm_prompts.py [hunk 1]
```diff
@@ -104,37 +104,38 @@ def get_supported_languages(cls) -> list[str]:
...
-    the question. Be succinct and do not include any unnecessary information.
-    Ignore any redacted and offensive words.
-    If no paraphrase is possible, respond with {PARAPHRASE_FAILED_MESSAGE}
-    """
-
-# ----  Question answering bot
-
-ANSWER_QUESTION_PROMPT = """
-    You are a high-performing question answering bot.
-    You support a maternal and child health chatbot.
```
Comment (suzinyou) [inline core_backend/app/configs/llm_prompts.py:128]: Removed this line

File: core_backend/app/routers/question_answer.py [hunk 2]
```diff
@@ -112,6 +115,23 @@ async def get_user_query_and_response(
     return user_query_db, user_query_refined, response
 
 
+def get_context_string_from_retrieved_contents(
```
Comment (suzinyou) [inline core_backend/app/routers/question_answer.py:118]: I created this so we can summarize multiple top contents!

File: core_backend/app/routers/manage_content.py [hunk 3]
```diff
@@ -48,7 +49,10 @@ async def create_content(
     """
 
     payload = _create_payload_for_qdrant_upsert(
-        content.content_title, content.content_text, content.content_metadata
+        content.content_title,
+        content.content_text,
+        content.content_metadata,
+        content_language=content.content_language,
```
Comment (amiraliemami) [inline core_backend/app/routers/manage_content.py:55]: Can you either add all variables names or remove `content_language=` for consistency? I always prefer writing the var names cos even if the order changes in a refactor the calls will still work

File: core_backend/app/routers/manage_content.py [hunk 4]
```diff
@@ -26,8 +27,8 @@ class QdrantPayload(BaseModel):
     # Ensure len("*{title}*\n\n{text}") <= 1600
     content_title: Annotated[str, StringConstraints(max_length=150)]
     content_text: Annotated[str, StringConstraints(max_length=1446)]
+    content_language: str = Language.ENGLISH.value
```
Comment (amiraliemami) [inline core_backend/app/routers/manage_content.py:30]: default should be UNKOWN?

File: core_backend/tests/api/test_manage_content.py [hunk 5]
```diff
@@ -303,6 +303,7 @@ def _search_qdrant_collection_by_id(self, content_id: uuid.UUID) -> List[Record]
         [
             ("title1", "content text 1", {}),
             ("title 1", "content text 2", {"meta_key": "meta_value"}),
+            ("title 1", "content text 2", {"meta_key": "meta_value"}),
         ],
```
Comment (amiraliemami) [inline core_backend/tests/api/test_manage_content.py:307]: What's the purpose of this repeat?

**Review by amiraliemami — APPROVED** [review]
Nice, runs via docker compose and all pytests pass - thanks! Left a frew comments. I was also thinking - isn't language a type of metadata? We have the metadata field in different functions but the language sits outside of that dict now as a separate entity. We should rethink that field maybe.

File: core_backend/tests/api/test_manage_content.py (hunk 5 above)
Comment (suzinyou) [reply core_backend/tests/api/test_manage_content.py:307]: It's a mistake :'( Thank you for catching!

File: core_backend/app/db/vector_db.py [hunk 6]
```diff
@@ -52,51 +56,71 @@ def create_qdrant_collection(collection_name: str, embeddings_dim: str) -> None:
...
 
 
 async def get_similar_content_async(
-    question: UserQueryBase,
+    question: UserQueryRefined,
     qdrant_client: QdrantClient,
     n_similar: int,
-    qdrant_collection_name: str = QDRANT_COLLECTION_NAME,
+    question_language: Language = Language.ENGLISH,
 ) -> Dict[int, UserQuerySearchResult]:
```
[the remark this answers is not in the document]
Comment (suzinyou) [reply core_backend/app/db/vector_db.py:83]: Thank you!!

**Comment by suzinyou** [comment]
> Nice, runs via docker compose and all pytests pass - thanks! Left a frew comments. I was also thinking - isn't language a type of metadata? We have the metadata field in different functions but the language sits outside of that dict now as a separate entity. We should rethink that field maybe.

Good point. Though Language will become a required field when we move onto multilanguage content DB.

**Review by amiraliemami — APPROVED** [review]
Thanks for addressing the comments! All tests pass and runs in prod mode too :)

**Pull request merged**